File: permissions/src/main/java/com/gabriel/permissions/ui/controller/UsersHttpController.java
package com.gabriel.permissions.ui.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.gabriel.permissions.infraestructure.provider.Auth0Provider;
import com.gabriel.permissions.ui.controller.request.CredentialsRequest;
import com.gabriel.permissions.ui.controller.request.RegisterRequest;
import com.gabriel.permissions.ui.controller.response.AuthenticationResponse;
import jakarta.servlet.http.HttpServletRequest;
import kong.unirest.core.HttpResponse;
import kong.unirest.core.Unirest;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.security.Principal;


@RestController
public class UsersHttpController {

    private String issuer;
    private String clientId;
    private String clientSecret;
    private String audience;
    private String scope;
    private String logoutRedirectUrl;
    private ObjectMapper objectMapper;

    private Auth0Provider auth0Provider;

    public UsersHttpController(
        @Value("${auth0.issuer}") String issuer,
        @Value("${auth0.clientId}") String clientId,
        @Value("${auth0.clientSecret}") String clientSecret,
        @Value("${auth0.audience}") String audience,
        @Value("${auth0.scope}") String scope,
        @Value("${auth0.logout_url}") String logoutRedirectUrl,
        ObjectMapper objectMapper,
        Auth0Provider auth0Provider
    ) {
        this.issuer = issuer;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.audience = audience;
        this.scope = scope;
        this.logoutRedirectUrl = logoutRedirectUrl;
        this.objectMapper = objectMapper;
        this.auth0Provider = auth0Provider;
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody CredentialsRequest credentialsRequest) {
        try {
            HttpResponse<String> response = Unirest.post("https://" + issuer + "/oauth/token")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .field("grant_type", "password")
                .field("client_id", clientId)
                .field("client_secret", clientSecret)
                .field("audience", audience)
                .field("scope", scope)
                .field("username", credentialsRequest.username())
                .field("password", credentialsRequest.password())
                .asString();

            if (response.isSuccess()) {
                AuthenticationResponse authenticationResponse = objectMapper.readValue(response.getBody(), AuthenticationResponse.class);
                return ResponseEntity.ok().body(authenticationResponse);
            } else {
                return ResponseEntity.status(response.getStatus()).body("Authentication failed");
            }
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("Server error during authentication");
        }
    }

    @PostMapping("/logout")
    public ResponseEntity<?> logout(HttpServletRequest request) {
        String logoutUrl = "https://" + issuer + "/v2/logout?client_id=" + clientId + "&returnTo=" + logoutRedirectUrl;
        return ResponseEntity.status(HttpStatus.FOUND).location(URI.create(logoutUrl)).build();
    }


    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody RegisterRequest registerRequest) {
        try {
            String token = auth0Provider.getManagementApiToken();
            HttpResponse<String> response = Unirest.post("https://" + issuer + "/dbconnections/signup")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .header("Authorization", "Bearer " + token)
                .field("client_id", clientId)
                .field("username", registerRequest.username())
                .field("password", registerRequest.password())
                .field("name", registerRequest.name())
                .field("email", registerRequest.email())
                .field("verify_email", "false")
                .field("connection", "Username-Password-Authentication")
                .asString();


            if (response.isSuccess()) {
                return ResponseEntity.ok("User registered successfully");
            } else {
                return ResponseEntity.status(response.getStatus()).body("Failed to create account: " + response.getBody());
            }
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("Server error during account creation");
        }
    }

    @DeleteMapping("/delete-account")
    public ResponseEntity<?> deleteAccount(Principal principal) {
        try {
            String token = auth0Provider.getManagementApiToken();
            HttpResponse<String> response = Unirest.delete("https://" + issuer + "/api/v2/users/" + encodeUserId(principal.getName()))
                .header("Authorization", "Bearer " + token)
                .asString();

            if (response.isSuccess()) {
                return ResponseEntity.ok("Account deleted");
            } else {
                return ResponseEntity.status(response.getStatus()).body("Failed to delete account: " + response.getBody());
            }
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("Server error during account deletion");
        }
    }

    private String encodeUserId(String userId) {
        try {
            return URLEncoder.encode(userId, StandardCharsets.UTF_8);
        } catch (Exception e) {
            throw new RuntimeException("Error encoding user ID", e);
        }
    }

}

File: permissions/src/main/kotlin/com/gabriel/permissions/ui/controller/UsersHttpController.kt
package com.gabriel.permissions.ui.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.gabriel.permissions.infraestructure.provider.Auth0Provider
import com.gabriel.permissions.ui.controller.request.CredentialsRequest
import com.gabriel.permissions.ui.controller.request.RegisterRequest
import com.gabriel.permissions.ui.controller.response.AuthenticationResponse
import kong.unirest.core.Unirest
import org.eclipse.microprofile.config.inject.ConfigProperty
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import jakarta.annotation.security.PermitAll
import jakarta.ws.rs.*
import jakarta.ws.rs.core.Context
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import jakarta.ws.rs.core.SecurityContext

@Path("/users")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
class UsersHttpController(
    @ConfigProperty(name = "auth0.issuer") private val issuer: String,
    @ConfigProperty(name = "auth0.client-id") private val clientId: String,
    @ConfigProperty(name = "auth0.client-secret") private val clientSecret: String,
    @ConfigProperty(name = "auth0.audience") private val audience: String,
    @ConfigProperty(name = "auth0.scope") private val scope: String,
    @ConfigProperty(name = "auth0.logout-url") private val logoutRedirectUrl: String,
    private val objectMapper: ObjectMapper,
    private val auth0Provider: Auth0Provider
) {

    @POST
    @Path("/login")
    @PermitAll
    fun login(credentialsRequest: CredentialsRequest): Response {
        return try {
            val response = Unirest.post("https://$issuer/oauth/token")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .field("grant_type", "password")
                .field("client_id", clientId)
                .field("client_secret", clientSecret)
                .field("audience", audience)
                .field("scope", scope)
                .field("username", credentialsRequest.username)
                .field("password", credentialsRequest.password)
                .asString()

            if (response.isSuccess) {
                val authenticationResponse = objectMapper.readValue(response.body, AuthenticationResponse::class.java)
                Response.ok(authenticationResponse).build()
            } else {
                Response.status(response.status).entity("Authentication failed").build()
            }
        } catch (e: Exception) {
            Response.serverError().entity("Server error during authentication").build()
        }
    }

    @POST
    @Path("/logout")
    @PermitAll
    fun logout(): Response {
        val logoutUrl = "https://$issuer/v2/logout?client_id=$clientId&returnTo=$logoutRedirectUrl"
        return Response.seeOther(java.net.URI.create(logoutUrl)).build()
    }

    @POST
    @Path("/register")
    @PermitAll
    fun register(registerRequest: RegisterRequest): Response {
        return try {
            val token = auth0Provider.getManagementApiToken()
            val response = Unirest.post("https://$issuer/dbconnections/signup")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .header("Authorization", "Bearer $token")
                .field("client_id", clientId)
                .field("username", registerRequest.username)
                .field("password", registerRequest.password)
                .field("name", registerRequest.name)
                .field("email", registerRequest.email)
                .field("verify_email", "false")
                .field("connection", "Username-Password-Authentication")
                .asString()

            if (response.isSuccess) {
                Response.ok("User registered successfully").build()
            } else {
                Response.status(response.status).entity("Failed to create account: ${response.body}").build()
            }
        } catch (e: Exception) {
            Response.serverError().entity("Server error during account creation").build()
        }
    }

    @DELETE
    @Path("/delete-account")
    fun deleteAccount(@Context securityContext: SecurityContext): Response {
        return try {
            val principalName = securityContext.userPrincipal?.name ?: return Response.status(Response.Status.UNAUTHORIZED).build()
            val token = auth0Provider.getManagementApiToken()
            val encodedUserId = URLEncoder.encode(principalName, StandardCharsets.UTF_8)
            val response = Unirest.delete("https://$issuer/api/v2/users/$encodedUserId")
                .header("Authorization", "Bearer $token")
                .asString()

            if (response.isSuccess) {
                Response.ok("Account deleted").build()
            } else {
                Response.status(response.status).entity("Failed to delete account: ${response.body}").build()
            }
        } catch (e: Exception) {
            Response.serverError().entity("Server error during account deletion").build()
        }
    }
}
